File: 0098_validate_binary_search_tree/is_valid_BST.java
public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    // Time: O(n)
    // Space: O(1)
    public boolean isValidBST(TreeNode root) {
        return helper(root, null, null);
    }
    
    private boolean helper(TreeNode p, Integer low, Integer high) {
        if (p == null)  {
            return true;
        }
        if ((low != null && p.val <= low) || (high != null && p.val >= high)) {
            return false;
        }
        if (!helper(p.left, low, p.val) || !helper(p.right, p.val, high)) {
            return false;
        }
        return true;
    }
}

File: 0098_validate_binary_search_tree/is_valid_BST.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def isValidBST(root: TreeNode) -> bool:
    """
    Time: O(n)
    Space: O(1)
    """
    return helper(root)
    

def helper(node: TreeNode, MIN = None, MAX = None):
    if not node:
        return True
    if (MAX != None and node.val >= MAX) or (MIN != None and node.val <= MIN):
        return False
    if not helper(node.left, MIN, node.val) or not helper(node.right, node.val, MAX):
        return False
    return True